### PR TITLE
check for coinbase maturity

### PR DESCRIPTION
**Pull request by mubarak23**
- [x] issue #108 
- [x] follows contribution [guide](https://github.com/keep-starknet-strange/raito/blob/main/CONTRIBUTING.md)

### CODE REVIEW

File: src/validation/transaction.cairo [hunk 1]
```diff
@@ -29,6 +29,22 @@ pub fn validate_transaction(
     //      - https://github.com/bitcoin/bitcoin/blob/master/src/consensus/tx_verify.cpp
     //      - https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp
 
+    //      - Coinbase is mature (if some input spends coinbase tx from the past)
+    let first_input = *tx.inputs[0];
+    if first_input.previous_output.is_coinbase {
```
Comment (maciejka) [inline src/validation/transaction.cairo:34]: Maturity check should be applied to all inputs.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: src/validation/transaction.cairo [hunk 2]
```diff
@@ -30,20 +30,28 @@ pub fn validate_transaction(
...
-        }
-    }
+    let mut errors = ArrayTrait::new();
+
+    for input in *tx
+        .inputs {
+            if *input.previous_output.is_coinbase {
+                let coinbase_block_height = *input.previous_output.block_height;
+                if block_height <= coinbase_block_height + 100 {
+                    errors
```
Comment (maciejka) [inline src/validation/transaction.cairo:40]: Why not break on first error?

**Review by maciejka — CHANGES_REQUESTED** [review]
@m-kus do we want to add is_coinbase to get_block.py script, or it will be superseded by your work on new testing infra?

File: src/validation/transaction.cairo (hunk 2 above)
Comment (mubarak23) [reply src/validation/transaction.cairo:40]: ok

Comment (mubarak23) [reply src/validation/transaction.cairo:40]: i have the following 

```
    if !errors.is_empty() {
        return Result::Err( *errors[0]
            );        
    },
```
    
    `return Result::Err( *errors[0] ` throw the following error 
    
```
            return Result::Err( *errors[0]
                            ^********^
note: Trait has no implementation in context: core::traits::Copy::<core::byte_array::ByteArray>.
```
@maciejka what could be wrong here,

Comment (mubarak23) [reply src/validation/transaction.cairo:40]: > i have the following
> 
> ```
>     if !errors.is_empty() {
>         return Result::Err( *errors[0]
>             );        
>     },
> ```
> 
> ```
> `return Result::Err( *errors[0] ` throw the following error 
> ```
> 
> ```
> error: Cannot desnap a non copyable type.
>  --> /Users/macbookpro/Documents/onlydust/raito/src/validation/transaction.cairo:58:29
>         return Result::Err( *errors[0]
>                             ^********^
> note: Trait has no implementation in context: core::traits::Copy::<core::byte_array::ByteArray>..
> ```
> 
> @maciejka what could be wrong here,

fix this, by using the clone methond on` errors[0]`

File: scripts/data/get_block.py [hunk 3]
```diff
@@ -65,6 +65,7 @@ def outpoint(prev_output):
                                     data: {tx_output({'vout': [prev_tx['vout'][prev_output['vout']]]})}
                                     block_height: {prev_block['height']}_u32,
                                     block_time: {prev_block['time']}_u32,
+                                    is_coinbase: {str(prev_output['vout'] == 0).lower()}
```
Comment (m-kus) [inline scripts/data/get_block.py:68]: It should be something like `prev_tx['vin'][0].get('coinbase') is not None`

Comment (maciejka) [reply scripts/data/get_block.py:68]: First transaction is not always the coinbase one?

Comment (mubarak23) [reply scripts/data/get_block.py:68]: what is right way to know the coinbase transaction

Comment (mubarak23) [reply scripts/data/get_block.py]: @maciejka  i was confuse on what do here, sow your last commit, everything is clear now

File: src/types/transaction.cairo [hunk 4]
```diff
@@ -97,6 +97,7 @@ pub struct OutPoint {
     /// containing this output was mined).
     /// Can be set to default if locktime feature (time relative) is disabled.
     pub block_time: u32,
+    pub is_coinbase: bool
```
Comment (m-kus) [inline src/types/transaction.cairo:100]: Pls add docstring

Comment (mubarak23) [reply src/types/transaction.cairo:100]: ok

File: src/validation/transaction.cairo [hunk 5]
```diff
@@ -29,6 +29,34 @@ pub fn validate_transaction(
     //      - https://github.com/bitcoin/bitcoin/blob/master/src/consensus/tx_verify.cpp
     //      - https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp
 
+    let mut maturity_result = Option::None;
+
+    // Coinbase maturity test
+    for input in *tx
```
Comment (m-kus) [inline src/validation/transaction.cairo:35]: Use break with value approach like here https://github.com/keep-starknet-strange/raito/blob/e3b413a6b9382e30eaa46d654272a434e159bdb0/src/validation/block.cairo#L19

Comment (mubarak23) [reply src/validation/transaction.cairo:35]: Did something like this 
```

   let maturity_result: Result<(), ByteArray> = loop {
        for input in *tx
            .inputs {
                if *input.previous_output.is_coinbase {
                    let coinbase_block_height = *input.previous_output.block_height;
                    if block_height <= coinbase_block_height + 100 {
                        let err = format!(
                                "[validate_transaction] coinbase input: ({}, {}) of tx: {} not mature (current height: {}, coinbase height: {})",
                                input.previous_output.txid,
                                input.previous_output.vout,
                                tx.txid(),
                                block_height,
                                coinbase_block_height
                            );
                        break Result::Err(
                            err
                        );
                    }
                }
            };
        break Result::Ok(());
    };

    maturity_result?;
```

am getting the following error on SCARB BUILD
```

error: Can only break with a value inside a `loop`.
 --> /Users/macbookpro/Documents/onlydust/raito/src/validation/transaction.cairo:47:25
                        break Result::Err(
```

@m-kus

**Review by maciejka — APPROVED** [review]

**Pull request merged**